test/solidity/oz/transparent-upgradeable-proxy/transparentProxy.js

```diff
@@ -0,0 +1,107 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+const chai = require('chai')
+const { expect } = require('chai')
+const chaiAsPromised = require("chai-as-promised")
+const { ethers } = require('hardhat')
+const Constants = require('../../../constants')
+
+chai.use(chaiAsPromised);
+
+describe('@solidityequiv3 Transparent Upgradeable Proxy', function () {
+  let contractProxy, contractBox
+  let owner, signer, proxyAdminAddress
+  before(async function () {
+    [owner, signer] = await ethers.getSigners()
+    const factoryBox = await ethers.getContractFactory(Constants.Contract.Box)
+    contractBox = await factoryBox.deploy()
+    await contractBox.deployed()
+
+    const factory = await ethers.getContractFactory(Constants.Contract.MyCustomTransparentUpgradeableProxy)
+    contractProxy = await factory.deploy(contractBox.address, owner.address, [])
+    const receipt = await contractProxy.deployTransaction.wait()
+
+    proxyAdminAddress = receipt.events[2].args.newAdmin
+  })
+
+  it('should verify it calls the correct contract and method via proxy', async function () {
+    const signedTx = await owner.sendTransaction({
+        to: contractProxy.address,
+        data: '0x6057361d0000000000000000000000000000000000000000000000000000000000000008',
+        gasLimit: 5000000
+    })
+    const receipt = await signedTx.wait()
+
+    expect(receipt.to).to.eq(contractProxy.address)
+    expect(receipt.from).to.eq(owner.address)
+    expect(receipt.logs[0].data).to.eq('0x0000000000000000000000000000000000000000000000000000000000000008')
+  })
+
+  it('should verify it can change the underlying contract', async function () {
+    const factoryBoxV2 = await ethers.getContractFactory(Constants.Contract.BoxV2)
+    contractBoxV2 = await factoryBoxV2.deploy()
+    await contractBoxV2.deployed()
+
+    const functionSelectorUpgradeAndCall = ethers.utils.keccak256(ethers.utils.toUtf8Bytes("upgradeAndCall(address,address,bytes)")).substring(0,10)
+    const abi = ethers.utils.defaultAbiCoder;
+    const encoded = abi.encode(["address","address","bytes"],[contractProxy.address, contractBoxV2.address, []])
+
+    const signedTx = await owner.sendTransaction({
+        to: proxyAdminAddress,
+        data: functionSelectorUpgradeAndCall + encoded.replace('0x', ''),
+        gasLimit: 5000000
+    })
+
+    const receipt = await signedTx.wait()
+
+    const topics = receipt.logs[0].topics
+    const eventUpgradedNameHashed = ethers.utils.keccak256(ethers.utils.toUtf8Bytes("Upgraded(address)"))
+    const newContractAddressEncoded = '0x000000000000000000000000' + contractBoxV2.address.replace('0x','')
+    expect(eventUpgradedNameHashed).to.eq(topics[0])
+    expect(newContractAddressEncoded.toLowerCase()).to.eq(topics[1])
+
+    const functionSelectorIncrement = ethers.utils.keccak256(ethers.utils.toUtf8Bytes("increment()")).substring(0,10)
```

Comment:
I see that this operation is repeating several times, maybe it's a good idea to take it out in utils ?